src/jace/stages.py

```diff
@@ -50,203 +51,225 @@
+        jit_options: api.JitOptions,
     ) -> None:
+        assert all(
+            param.default is param.empty for param in inspect.signature(fun).parameters.values()
+        )
         super().__init__()
-        # We have to shallow copy both the translator and the jit options.
-        #  This prevents that any modifications affect `self`.
-        #  Shallow is enough since the translators themselves are immutable.
         self._primitive_translators = {**primitive_translators}
-        # TODO(phimuell): Do we need to deepcopy the options?
         self._jit_options = {**jit_options}
         self._fun = fun
 
-    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+    def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> _RetrunType:
         """
         Executes the wrapped function, lowering and compiling as needed in one step.
 
-        The arguments passed to this function are the same as the wrapped function uses.
+        This function will lower and compile in one go. The function accepts the same
+        arguments as the original computation and the return value is unflattened.
+
+        Note:
+            This function is also aware if a Jax tracing is going on. In this
+            case, it will forward the computation.
+            Currently, this function ignores the value of `jax.disable_jit()`.
         """
-        # If we are inside a traced context, then we forward the call to the wrapped
-        #  function. This ensures that JaCe is composable with Jax.
         if util.is_tracing_ongoing(*args, **kwargs):
             return self._fun(*args, **kwargs)
 
         lowered = self.lower(*args, **kwargs)
         compiled = lowered.compile()
+        # TODO(phimuell): Filter out static arguments
         return compiled(*args, **kwargs)
 
     @tcache.cached_transition
-    def lower(self, *args: Any, **kwargs: Any) -> JaCeLowered:
+    def lower(self, *args: _P.args, **kwargs: _P.kwargs) -> JaCeLowered[_RetrunType]:
         """
-        Lower this function explicitly for the given arguments.
+        Lower the wrapped computation for the given arguments.
+
+        Performs the first two steps of the AOT steps described above, i.e. trace the
+        wrapped function with the given arguments and stage it out to a Jaxpr. Then
+        translate it to an SDFG. The result is encapsulated inside a `JaCeLowered`
+        object that can later be compiled.
 
-        Performs the first two steps of the AOT steps described above, i.e.
-        trace the wrapped function with the given arguments and stage it out
-        to a Jaxpr. Then translate it to SDFG. The result is encapsulated
-        inside a `JaCeLowered` object which can later be compiled.
+        It should be noted that the current lowering process will hard code the strides
+        and the storage location of the input inside the SDFG. Thus if the SDFG is
+        lowered with arrays in C order, calling the compiled SDFG with FORTRAN order
```

Comment:
No, as DaCe is concerned an array argument is a pointer.


FYI: It also does not consider the datatype, this is the actual reason for why we trace with enabled `x64` mode.
In addition if you have a dynamic size you also have to pass this size as an additional argument in the call.
The only thing it complains about if you confuses scalars with arrays or GPU vs. CPU memory.

In JaCe as long as you do not use the AOT mode you are safe because the cache will take care of it.